src/EnergyPlus/WaterThermalTanks.cc

```diff
@@ -1042,25 +1045,30 @@ namespace WaterThermalTanks {
             // Initialize the offsets to zero
             nAlphaOffset = 0;
             nNumericOffset = 0;
+
             if (HPWaterHeaterNum <= NumPumpedCondenser) {
                 // Pumped Condenser
                 DataIPShortCuts::cCurrentModuleObject = cHPWHPumpedCondenser;
                 HPWH.TypeNum = DataPlant::TypeOf_HeatPumpWtrHeaterPumped;
                 nNumPossibleAlphaArgs = 29;
                 nNumPossibleNumericArgs = 9;
+                // Actual index of Pumped type
+                HPWaterHeaterNumOfSpecificType = HPWaterHeaterNum;
             } else {
                 // Wrapped Condenser
                 DataIPShortCuts::cCurrentModuleObject = cHPWHWrappedCondenser;
                 HPWH.TypeNum = DataPlant::TypeOf_HeatPumpWtrHeaterWrapped;
                 nNumPossibleAlphaArgs = 27;
                 nNumPossibleNumericArgs = 10;
+                // Actual index of Wrapped type
+                HPWaterHeaterNumOfSpecificType = HPWaterHeaterNum - NumPumpedCondenser;
```

Comment:
This is the actual fix basically.

Container `HPWaterHeater` stores both HPWH:Pumped and HPWH:Wrapped in that order. Before fix if you had 1 HPWH:Pumped and one HPWH:Wrapped it would call getObjectItem for the wrapped one passing an index of 2, which doesn't exist in InputProcessor since there's only one object of type HPWH:WrappedCondenser, so it would crash hard.

Reply:
@jmarrec thanks for the fix. Having just worked through the refactor on this file, I can attest that there are several things like this in here that need to get cleaned up.